src/main/java/uk/nhs/hee/tis/trainee/forms/migration/AddPartialDeleteMetadataToS3.java

```
@@ -0,0 +1,101 @@
+
+/**
+ * Add partial delete related metadata to existing forms on S3.
+ */
+@Slf4j
+@ChangeUnit(id = "AddPartialDeleteMetadataToS3", order = "7")
+public class AddPartialDeleteMetadataToS3 {
+
+  private final AmazonS3 amazonS3;
+  private final String bucketName;
+
+  private static final String FIXED_FIELDS =
+      "id,traineeTisId,lifecycleState,submissionDate,lastModifiedDate";
+
+  /**
+   * Add partial delete related metadata to existing forms on S3.
+   */
+  public AddPartialDeleteMetadataToS3(AmazonS3 amazonS3, Environment env) {
+    this.amazonS3 = amazonS3;
+    this.bucketName = env.getProperty("application.file-store.bucket");
+  }
+
+  /**
+   * Find all object on S3 and add metadata.
+   */
+  @Execution
+  public void migrate() {
+    log.info("Starting migration to add partial delete related metadata to existing forms on S3.");
+
+    try {
+      List<S3ObjectSummary> keyList = new ArrayList<>();
+      var objects = amazonS3.listObjects(bucketName);
+      keyList.addAll(objects.getObjectSummaries());
+
+      while (objects.isTruncated()) {
+        objects = amazonS3.listNextBatchOfObjects(objects);
+        keyList.addAll(objects.getObjectSummaries());
+      }
+      log.info("Updating {} objects in the bucket {}", keyList.size(), bucketName);
+
+      for (var obj : keyList) {
+        final var object = amazonS3.getObject(bucketName, obj.getKey());
+
+        var metadata = object.getObjectMetadata();
+        metadata.addUserMetadata("deletetype", DeleteType.PARTIAL.name());
+        metadata.addUserMetadata("fixedfields", FIXED_FIELDS);
+
+        final var request = new PutObjectRequest(
+            bucketName, obj.getKey(), object.getObjectContent(), metadata);
+        amazonS3.putObject(request);
+      }
+
+      log.info("Finish migration");
+    }
+    catch (Exception e) {
+      log.error("Fail to add metadata to existing forms in bucket " + bucketName + ": ", e);
```

Given there zero downtime implications for this migrator, I'd maybe let the exception be thrown to stop the migrator.

Already migrated forms should not be reverted, but it gives us a clear indication that we need to fix and re-run rather than having to manually fix afterwards with no option to rerun the migration (without DB fiddling). Combined with the metadata check above we should have a fairly streamlined fail and retry process.